luda-editor/new-server/bptree/src/id_set/wal/mod.rs

```diff
@@ -0,0 +1,314 @@
+            .await?;
+
+        let shadow_file = OpenOptions::new()
+            .create(true)
+            .read(true)
+            .write(true)
+            .truncate(false)
+            .open(path.with_extension("shadow"))
+            .await?;
+
+        let (wal_read_fd, wal_write_fd) = split_file(wal_file.into_std().await);
+        let (shadow_read_fd, mut shadow_write_fd) = split_file(shadow_file.into_std().await);
+
+        let (tx, rx) = mpsc::unbounded_channel();
+        let (executer_close_tx, executer_close_rx) = oneshot::channel();
+
+        let mut this = Self {
+            wal_write_fd,
+            read_offset: ReadOffset::new(),
+            track_flag: true,
+            write_offset: 0,
+            written: 0,
+            tx,
+            executer_close_rx,
+        };
+
+        if file_write_fd.len()? == 0 {
+            this.write_wal(Init)?;
+            this.wal_write_fd.fsync()?;
+        }
+
+        let wal_file_len = this.wal_write_fd.len()?;
+
+        let mut read_offset: usize = 0;
+        while read_offset < wal_file_len {
+            match execute_one(&wal_read_fd, read_offset, &mut shadow_write_fd).await {
+                Ok(new_read_offset) => {
+                    read_offset = new_read_offset;
+                }
+                Err(err) => {
+                    if err.is_corrupted() {
+                        break;
+                    }
+                    match err {
+                        WalError::Executor(execute_error) => match execute_error {
+                            ExecuteError::Checksum { .. } => unreachable!(),
+                            ExecuteError::WrongBodyType { .. } => unreachable!(),
+                            ExecuteError::WrongBodySize { .. } => unreachable!(),
+                        },
+                        WalError::Io(error) => {
+                            return Err(error);
+                        }
+                        WalError::ExecutorDown => unreachable!(),
+                    }
+                }
+            };
+        }
+        if wal_file_len > 0 {
+            file_write_fd.copy_from(&shadow_read_fd)?;
+
+            this.wal_write_fd.set_len(0)?;
+            this.wal_write_fd.fsync()?;
+            this.write_offset = 0;
+            this.written = 0;
+        }
+
+        Executor::new(
+            wal_read_fd,
+            shadow_write_fd,
+            rx,
+            this.read_offset.clone(),
+            executer_close_tx,
+        )
+        .start();
+
+        Ok(this)
+    }
+
+    pub(crate) fn update_pages(&mut self, pages: &BTreeMap<PageOffset, Page>) -> Result<()> {
+        self.start_write()?;
+
+        if pages.is_empty() {
+            return Ok(());
+        }
+
+        for (offset, page) in pages {
+            let put_page = PutPage {
+                page_offset: *offset,
+                page: page.clone(),
+            };
+
+            self.write_wal(put_page)?;
+        }
+
+        self.sync_all()?;
```

Comment:
This part is incorrect. On the side reading from the WAL, there is no way to determine whether only part of the data was written to the WAL due to a system crash or other issue. There should not be a checksum for a single page write, such as with `PutPage`. Instead, all Page changes should be written at once, atomically, with a multiple-page write through `PutPages`.